src/eterna/mode/Mol3DGate.ts

```diff
@@ -0,0 +1,291 @@
+      return navigator.userAgent.match(/Android/i);
+    },
+    BlackBerry: function BlackBerry() {
+      return navigator.userAgent.match(/BlackBerry/i);
+    },
+    iOS: function iOS() {
+      return navigator.userAgent.match(/iPhone|iPad|iPod/i);
+    },
+    Opera: function Opera() {
+      return navigator.userAgent.match(/Opera Mini/i);
+    },
+    Windows: function Windows() {
+      return (
+        navigator.userAgent.match(/IEMobile/i) ||
+        navigator.userAgent.match(/WPDesktop/i)
+      );
+    },
+    any: function any() {
+      return (
+        checker.Android() ||
+        checker.BlackBerry() ||
+        checker.iOS() ||
+        checker.Opera() ||
+        checker.Windows()
+      );
+    },
+  };
+  return !!checker.any();
+};
+
+const DectectDevice = () => {
+  isMobile();
+};
+
+interface PixiCifCheckerCallback {
+  (component: NGL.Structure | null): void;
+}
+
+export default class Mol3DGate {
+  public stage: NGL.Stage;
+  public component: NGL.Component | null;
+  protected backboneElement: NGL.RepresentationElement | null;
+  protected baseElement: NGL.RepresentationElement | null;
+  protected ballstickElement: NGL.RepresentationElement | null;
+  public static scope: Mol3DGate;
+  protected poseMode: GameMode;
+  private readonly secStruct: string;
+  protected colorChangeMap = new Map();
+  protected hoverdInfo = { index: -1, color: 0, outColor: 0 };
+  public bShowAnnotations: boolean = true;
+  protected threeView: ThreeView;
+  protected _3DFilePath: string | File | Blob = "";
+
+  protected myColorScheme: string;
+
+  constructor(
+    filePath: string | File | Blob,
+    container: HTMLElement,
+    threeView: ThreeView,
+    callback: NGL.PixiRenderCallback,
+    poseMode: GameMode,
+    _secStruct: string
+  ) {
+    Mol3DGate.scope = this;
+    this._3DFilePath = filePath;
+    this.poseMode = poseMode;
+    this.secStruct = _secStruct;
+    this.threeView = threeView;
+
+    const scheme = function xxx(this: NGL.Colormaker) {
+      this.atomColor = function yyy(atom: NGL.AtomProxy): number {
+        const colorArray = [0x3183c0, 0xaa1c20, 0xffff00, 0x1b7b3d];
+        let color = 0;
+        let atomResname = atom.resname;
+        Mol3DGate.scope.colorChangeMap.forEach((name, num) => {
+          if (atom.resno === num) {
+            atomResname = name;
+          }
+        });
+
+        switch (atomResname) {
+          case "U":
+            color = colorArray[0];
+            break;
+          case "G":
+            color = colorArray[1];
+            break;
+          case "A":
+            color = colorArray[2];
+            break;
+          case "C":
+            color = colorArray[3];
+            break;
+          default:
+            break;
+        }
+
+        return color;
+      };
+    };
+    this.myColorScheme = NGL.ColormakerRegistry.addScheme(
+      scheme,
+      "myColorScheme"
+    );
+
+    this.stage = new NGL.Stage(container, { mousePreset: "eterna" }, callback);
```

Comment:
One additional thing I noticed - you never seem to be cleaning up the NGL stuff. Namely, if we transition to a new GameMode, while the mode goes out of scope, NGL's resources probably aren't freed. The stage should be constructed within a GameObject that is added as a child of the mode. That way you can instantiate it within its added() method (ie, at creation time) and then you can call stage.dispose() in the dispose() method. I don't know if there's any other NGL-related objects (or anything else related to 3D rendering) you need to handle this way as well.

FWIW: This point goes back to the need to clarify what goes on Mol3DGate vs ThreeView. Probably makes more sense for the stage to be part of the ThreeView.

Reply:
Fixed. Now we call the dispose function of NGL in GameMode